### PR TITLE
Expand the discussion of external IP connectivity.

### LINKED ISSUE #4824: Document external routing into the network
Discussed at the face to face, but need to document in networking.md the external to internal solution space.

### PR DESCRIPTION
**Pull request by brendandburns**
Closes https://github.com/GoogleCloudPlatform/kubernetes/issues/4824

@thockin @smarterclayton 

I don't remember all of the details of the discussion, so it's possible that I forgot to add some things, please let me know if there are further things you'd like added.

--brendan

### CODE REVIEW

File: docs/design/networking.md [hunk 1]
```diff
@@ -97,11 +99,18 @@ So we end up with 3 cases:
...
 
-3. Internet -> Container. This also has to go through the primary host IP and also has 2 levels of NAT, ideally. However, the path currently is a proxy with (External Host IP -> Internal Host IP -> Docker) -> (Docker -> Container IP). Once [issue #15](https://github.com/GoogleCloudPlatform/kubernetes/issues/15) is closed, it should be External Host IP -> Internal Host IP -> Container IP. But to get that second arrow we have to set up the port forwarding iptables rules per mapped port.
+3. Internet -> Container. This also has to go through the primary host IP and has 1 level of NAT, ideally. However, the path currently is a proxy with (External Host IP -> Internal Host IP -> Docker) -> (Docker -> Container IP). Once [issue #15](https://github.com/GoogleCloudPlatform/kubernetes/issues/15) is closed, it should be External Host IP -> Container IP. The way that this will work is that each
+container that wants to be present on the external internet will specify an external ip address.  The IP address will be trapped via IPTables on the host node and re-written to the container IP address.
```
Comment (rajatchopra) [inline docs/design/networking.md]: Who will manage the iptables? Can a network-provider be free to not use iptables too, in lieu of its own solution. May not even need any extra work with an ipvlan based solution, e.g.

**Comment by smarterclayton** [comment]
This is LGTM - clear summary.

**Comment by smarterclayton** [comment]
Will defer to Tim for closure though

File: docs/design/networking.md (hunk 1 above)
Comment (brendandburns) [inline docs/design/networking.md]: the iptables is purely for external ip -> internal ip rewriting.  Even if you have ipvlan, I don't think that you can map the external IP to the pod without iptables, unless you somehow make the bridge that the container allocates addresses out of contain external IP addresses, but that seems pretty tricky (and dangerous) you're always going to want something that NATs the packet from external to internal IP.  I suppose you could do that at the edge instead, if you want to...

**Comment by brendandburns** [comment]
once again, I'm impressed my docs-only changes broke both travis and shippable...

**Comment by brendandburns** [comment]
@thockin I can haz an LGTM?

File: docs/design/networking.md [hunk 2]
```diff
@@ -97,11 +99,18 @@ So we end up with 3 cases:
 
 1. Container -> Container or Container <-> VM. These should use 10. addresses directly and there should be no NAT.
 
-2. Container -> Internet. These have to get mapped to the primary host IP so that GCE knows how to egress that traffic. There is actually 2 layers of NAT here: Container IP -> Internal Host IP -> External Host IP. The first level happens in the guest with IP tables and the second happens as part of GCE networking. The first one (Container IP -> internal host IP) does dynamic port allocation while the second maps ports 1:1.
+2. Container -> Internet. These have to get mapped to the primary host IP so that the host node knows how to egress that traffic. In most networking setups there are actually 2 layers of NAT here: Container IP -> Internal Host IP -> External Host IP. The first level happens in the guest with IP tables and the second happens as part of egress past a firewall to the broader internet.  On cloud providers, this is usually provided by the cloud virtual network.  On physical providers, this is provided by some sort of gateway router.. The first one (Container IP -> internal host IP) does dynamic port allocation while the second maps ports 1:1.
```
Comment (thockin) [inline docs/design/networking.md]: I'm not going to pick nits about being too GCE-centric - this doc needs an overhaul from that perspective.

"router.." - double dot

I think the last sentence here is more confusing than helpful.  Move to strike the whole sentence.  Specifically, outbound ports almost never matter, and I have not nearly enough breadth to endorse any statement of how these things are usually done outside of the core linux kernel implementation.

File: docs/design/networking.md (hunk 1 above)
Comment (thockin) [inline docs/design/networking.md]: I'm a little confused about the future-tense here.   Are you describing host ports (which is what #15 is about) or kube-proxy?  We shouldn't spend too much time on host ports, and we shouldn't be adding features in that space.

File: docs/design/networking.md (hunk 2 above)
Comment (thockin) [inline docs/design/networking.md]: I'm sticking with this change request

File: docs/design/networking.md (hunk 1 above)
Comment (thockin) [inline docs/design/networking.md]: disregard my comment this is fine.

**Comment by thockin** [comment]
LGTM % one change request, feel free to self-merge

File: docs/design/networking.md (hunk 2 above)
Comment (brendandburns) [inline docs/design/networking.md]: done. and done.

**Comment by brendandburns** [comment]
Comments addressed, will self-merge on green.

**Comment by thockin** [comment]
LGTM

**Comment by brendandburns** [comment]
Merging, given that Shippable is a flake and travis is running sloooooow

**Pull request merged**